Introduce implicit integer precision access operator for View

crtrott opened the pull request:
Addressing the performance issue in View by reintroducing a view access operator that uses the common type precision for mapping calculations.

crtrott committed:
Introduce implicit integer precision access operator for View

Signed-off-by: Christian Trott <crtrott@sandia.gov>

JBludau commented:
https://github.com/kokkos/kokkos/issues/8430

tcclevenger commented on core/src/Kokkos_View.hpp:
```diff
@@ -400,6 +404,107 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
+    using idx_type     = std::common_type_t<IndexOffsets...>;
+    idx_type indices[] = {static_cast<idx_type>(index_offsets)...};
+    // self-recursive fold trick from
+    // https://github.com/llvm/llvm-project/blob/96e1914aa2e6d8966acbfbe2f4d184201f1aa318/libcxx/include/mdspan/layout_left.h#L144
```
Link in the comment is not correct
```suggestion
    // https://github.com/llvm/llvm-project/blob/96e1914aa2e6d8966acbfbe2f4d184201f1aa318/libcxx/include/__mdspan/layout_left.h#L144
```

tcclevenger commented on core/src/Kokkos_View.hpp:
```diff
@@ -400,6 +404,107 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
+      std::index_sequence<Ranks...>, IndexOffsets... index_offsets) const {
+    using idx_type = std::common_type_t<IndexOffsets...>;
+    // self-recursive fold trick from
+    // https://github.com/llvm/llvm-project/blob/4d9771741d40cc9cfcccb6b033f43689d36b705a/libcxx/include/mdspan/layout_right.h#L141
```
Same as above
```suggestion
    // https://github.com/llvm/llvm-project/blob/4d9771741d40cc9cfcccb6b033f43689d36b705a/libcxx/include/__mdspan/layout_right.h#L141
```

JBludau commented:
@crtrott is this basically the impl of `mdspan` modified to use whatever indextype?

cwpearson commented on core/src/Kokkos_View.hpp:
```diff
@@ -204,6 +208,7 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
   // typedefs from BasicView
   using typename base_t::mdspan_type;
   using reference_type = typename base_t::reference;
+  using typename base_t::data_handle_type;
```
newly public - intentional?

crtrott replied:
yes needed by the bounds check macro.

crtrott replied:
and I mean long time this should be there anyway.

cwpearson commented on core/src/View/Kokkos_BasicView.hpp:
```diff
@@ -677,7 +677,8 @@ class BasicView {
                         m_map(static_cast<index_type>(std::move(indices))...));
   }
 
-#undef KOKKOS_IMPL_BASICVIEW_OPERATOR_VERIFY
```
What's behind this change?

crtrott replied:
As the comment says it's needed for the workaround. The Macro is now used inside Kokkos_View.hpp.

masterleinad replied:
I'd prefer to duplicate the macro definition instead of relying on inclusion order (and never undefining the macro even though it's clearly limited to these two files). An alternative would be moving it to a separate file that is included by both `Kokkos_BasicView.hpp` and `Kokkos_View.hpp` (but we still wouldn't `undef` the macro then).

crtrott commented:
> @crtrott is this basically the impl of `mdspan` modified to use whatever indextype?

Yes

crtrott commented:
The HIP jobs with 6.2.1 segfault with this variant. 6.4.0 works.

masterleinad reviewed (commented):
So `BasicView` has `operator()` and `operator[]` and this pull request introduces `operator()` in `View` with the same constraints and shadows the base class operator.
In other words, we get a better implementation for `operator()` but not for operator`[]` and `BasicView` `operator()` is unused now. Also `accesss` is using the base class operator.

Why exactly does the new operator need to live in `View`?

crtrott commented:
> So `BasicView` has `operator()` and `operator[]` and this pull request introduces `operator()` in `View` with the same constraints and shadows the base class operator. In other words, we get a better implementation for `operator()` but not for operator`[]` and `BasicView` `operator()` is unused now. Also `accesss` is using the base class operator.
> 
> Why exactly does the new operator need to live in `View`?

Valid point about access, I might need to fix that as for BasicView: for BasicView the index_type is explicitly specified in the template arguments, so this kind of shortcut here doesn't make sense. Furthermore, as much as possible I want to contain all the gnarly "get-rid-of-this-later" pieces of View in View, and keep them away from BasicView. The operator is right now used for "non-standard" layouts (which isn't really a thing yet admittedly) and also for ROCM 6.2/3. Furthermore, I am planning to introduce the alternative way of specifying View types (using mdspan template arguments) in which case these short cut operators would be disabled too.

dalg24 commented on core/src/Kokkos_View.hpp:
```diff
@@ -400,6 +405,106 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
   }
 #endif
 
+  // The following are shortcuts to allow implicit integer precision
```
Wondering if what "implicit" means here will be obvious enough to the reader or if we should clarify (something like `implicit (i.e. bla)`) that it means computing the offset in whatever integer type the indices arguments were passed.

And below for explicit, we could say we use 64-bit unsigned integers or just std::size_t

masterleinad replied:
Doesn't that mean we would overflow when accessing an allocation with span larger than `INT_MAX` using only `int` indices?

crtrott replied:
yes that is current behavior.

masterleinad replied:
AFAICT, with the legacy View implementation we multiplied by `m_dim.N*` in the index calculation and those are all of type `size_t` (for rank > 2) which means we would not overflow, see https://github.com/kokkos/kokkos/blob/3f21b30771c2a7a2f90dc3f62075b3cccbc023a5/core/src/View/Kokkos_ViewMapping.hpp#L832-L848 and https://github.com/kokkos/kokkos/blob/3f21b30771c2a7a2f90dc3f62075b3cccbc023a5/core/src/View/Kokkos_ViewDataAnalysis.hpp#L44-L68.

masterleinad replied:
We figured out that the legacy View implementation was indeed doing the index calculation in the access integer type for the most part (`m_dim.N*` are actually `unsigned int` for the dynamic extents and `m_stride` is always `size_type`). `[Basic]View::stride` still returns `size_t`.

dalg24 commented on core/src/Kokkos_View.hpp:
```diff
@@ -400,6 +405,106 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
+  // ROCM: The below code segfaults the compiler with ROCM 6.3 and ROCM 6.2
+  // We will simply avoid the performance optimization code path for those.
+#if !defined(HIP_VERSION) || HIP_VERSION_MAJOR > 6 || HIP_VERSION_MINOR > 3
+  KOKKOS_FUNCTION constexpr auto compute_offset(std::index_sequence<>) const {
```
I would add comments saying "rank 0, rank 1, rank 2+" or similar to help the reader

crtrott replied:
done

dalg24 commented on core/src/Kokkos_View.hpp:
```diff
@@ -400,6 +405,100 @@
+      std::index_sequence<Rank>, IndexOffset index_offset) const {
+    if constexpr (std::is_same_v<typename base_t::layout_type,
+                                 Kokkos::layout_stride>)
+      return index_offset * static_cast<IndexOffset>(m_map.stride(0));
```
Please make sure you fix this

dalg24 commented on core/src/Kokkos_View.hpp:
```diff
@@ -400,6 +405,106 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
+    return 0;
+  }
+
+  template <size_t Rank, class IndexOffset>
+  KOKKOS_FUNCTION constexpr auto compute_offset(
+      std::index_sequence<Rank>, IndexOffset index_offset) const {
```
```suggestion
  template <class IndexOffset>
  KOKKOS_FUNCTION constexpr auto compute_offset(
      std::index_sequence<0>, IndexOffset index_offset) const {
```

crtrott replied:
done

dalg24 commented on core/src/Kokkos_View.hpp:
```diff
@@ -400,6 +405,106 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
+
+  // CUDA: using requires clauses instead of if constexpr ran into issues
+  // with CUDA 12.2 + GCC 11.5, hence the if constexpr approach
+  template <size_t... Ranks, class... IndexOffsets>
```
I think `Ranks...` is not a good name choice because it can be somewhat confusing.
You could call it just `I...` that convey that you use it for indexing.

crtrott replied:
done

dalg24 commented on core/src/Kokkos_View.hpp:
```diff
@@ -400,6 +405,106 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
+  template <class... OtherIndexTypes>
+    requires(
+        (std::is_convertible_v<OtherIndexTypes, index_type> && ...) &&
+        (std::is_nothrow_constructible_v<index_type, OtherIndexTypes> && ...) &&
```
Isn't `index_type` always a regular integer type?
Why is convertible not enough?

crtrott replied:
this is the mdspan requirements no reason to differ from that.

crtrott replied:
and yes index_type is always regular, that doesn't mean that its no_throw_constructible from OtherIndexType if OtherIndexType is a class with a conversion operator that can throw for example.

dalg24 commented on core/src/Kokkos_View.hpp:
```diff
@@ -400,6 +405,106 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
+      OtherIndexTypes... indices) const {
+    KOKKOS_IMPL_BASICVIEW_OPERATOR_VERIFY(indices...);
+    return m_acc.access(m_ptr,
+                        m_map(static_cast<index_type>(std::move(indices))...));
```
Why are you moving these indices?
If there is no good reason to depart from the other overload, please use the same name for the trailing parameter pack.
Why are you casting?

crtrott replied:
That is what mdspan standard you are supposed to do.

dalg24 commented on core/src/View/Kokkos_BasicView.hpp:
```diff
@@ -677,7 +677,8 @@ class BasicView {
 
-#undef KOKKOS_IMPL_BASICVIEW_OPERATOR_VERIFY
+  // We need this for the workaround with integer promotion in View
+  // #undef KOKKOS_IMPL_BASICVIEW_OPERATOR_VERIFY
```
With this PR, when would the `BasicView::operator()` be called?

crtrott replied:
With ROCM, potentially with other layouts.

crtrott replied:
And I want to introduce next something that allows index_type specification for View in which case this new operator will not be enabled.

dalg24 reviewed (commented):
What about View::operator[] ?

masterleinad reviewed (dismissed):
We should make sure `operator[]` and `access` use the newly introduced call operator.

crtrott commented:
> What about View::operator[] ?

Not needed it anyway does 64bit for 1D, and in C++23 I don't want to introduce the shitty overflow semantics we have right now. I want to get rid of that behavior in the medium term anyway.

crtrott commented:
> We should make sure `operator[]` and `access` use the newly introduced call operator.

I do not want to introduce the shitty behavior in places where we don't absolutely have too. We may need to for access based on intrepid2 testing but lets cross that bridge when we get there.

masterleinad commented on core/src/Kokkos_View.hpp:
```diff
@@ -400,6 +405,108 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
+  // in offset calculations - meaning index calculation happens in the common
+  // type of the used indices. If and when these are not needed, the data access
+  // operator should come from BaseView only. BaseView will not support this
+  // directly, since BaseView anyway requires you to be explicit about the
```
```suggestion
  // operator should come from BasicView only. BasicView will not support this
  // directly, since BasicView anyway requires you to be explicit about the
```

crtrott replied:
done

crtrott commented:
@masterleinad you are wrong: 
```c++
  using Traits = Kokkos::ViewTraits<int****>;
  printf("%s\n",typeid(decltype(Kokkos::Impl::ViewOffset<typename Traits::dimension,
                                 typename Traits::array_layout, void>().m_dim.N0)).name());
 ```
prints unsigned int.

masterleinad commented:
> prints unsigned int.

https://godbolt.org/z/9YEec8z5n shows that we get `unsigned` for the first 4 dimensions and `size_t` otherwise.

crtrott commented:
> > prints unsigned int.
> 
> https://godbolt.org/z/9YEec8z5n shows that we get `unsigned` for the first 4 dimensions and `size_t` otherwise.

Yeah but the other 4 are not part of the View. dimensions (they are static values which are unused in the index calculation for a rank-4 view). So the actual calculation is done in unsigned int except for the multiplication with m_stride.

masterleinad commented:
The benchmark in https://github.com/kokkos/kokkos/pull/8490 shows on `Aurora`
```
========================================
Multidimensional View Index Benchmark
========================================
Concurrency:     1835008
Iterations:      100000
Backend:         SYCL
Team size:       64
League size:     28672
========================================
warming up int32_t/int64_t...
done
Running int32_t test...
done (output[0]=-1.8e+06)
Running int64_t test...
done (output[0]=-1.8e+06)
========================================
int32_t time:    19.317259 s
int64_t time:    9.662909 s
int32_t speedup: 0.50x
========================================
```
with this pull request whereas `int32_t` and `int64_t` were on par before (`develop`):
```
========================================
Multidimensional View Index Benchmark
========================================
Concurrency:     1835008
Iterations:      100000
Backend:         SYCL
Team size:       32
League size:     57344
========================================
warming up int32_t/int64_t...
done
Running int32_t test...
done (output[0]=-1.8e+06)
Running int64_t test...
done (output[0]=-1.8e+06)
========================================
int32_t time:    10.619829 s
int64_t time:    11.792611 s
int32_t speedup: 1.11x
========================================
```

masterleinad commented:
The lagcy View implementation shows
```
========================================
Multidimensional View Index Benchmark
========================================
Concurrency:     1835008
Iterations:      100000
Backend:         SYCL
Team size:       32
League size:     57344
========================================
warming up int32_t/int64_t...
done
Running int32_t test...
done (output[0]=-1.8e+06)
Running int64_t test...
done (output[0]=-1.8e+06)
========================================
int32_t time:    9.734209 s
int64_t time:    12.272254 s
int32_t speedup: 1.26x
========================================
```

masterleinad commented:
Using unsigned types doesn't really make any difference.

JBludau reviewed (dismissed):
I guess with the long term plan decided (https://github.com/kokkos/kokkos/pull/8476/files#r2395370069) this is a reasonable workaround for now.
There could be a fixme comment for later reference but since this is so major I hope we don't forget

masterleinad commented:
> Wait I read the perf results in https://github.com/kokkos/kokkos/pull/8476#issuecomment-3391378199 exactly wrong

Yeah, it's quite strange.

crtrott commented:
I don't see any problem with 2025.1

crtrott committed:
Adding a special workaround for SYCL

Daniel reported issues which I couldn't reproduce, but maybe this works
around them.

crtrott committed:
Revert "Adding a special workaround for SYCL"

This reverts commit 2ccc5690eb42daa457351faaf5a3f4864001cf3e.

crtrott commented:
I reverted that attempt to address the SYCL thing again.

stanmoore1 commented:
Here is my benchmarking for the LAMMPS SNAP machine learning potential on Aurora, measured in timesteps/second, so higher is better:
 | Kokkos Version | VIEW_LEGACY=on | VIEW_LEGACY=off |
 | :------- | :------: | -------: |
 | **4.7.1**    | 38.2   | 34.8   |
 | **PR #8476 2ccc5690eb42daa457351faaf5a3f4864001cf3e**   | 38.2   | 41.2   |
 
 Somehow this PR not only fixed the performance regression, but also made it even faster.

dalg24 commented on core/src/Kokkos_View.hpp:
```diff
@@ -397,6 +402,112 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
+         Kokkos::Impl::IsLayoutRightPadded<
+             typename base_t::layout_type>::value ||
+         std::is_same_v<typename base_t::layout_type, Kokkos::layout_stride>))
+  KOKKOS_FUNCTION constexpr base_t::reference operator()(
```
Seen in the CUDA-11.8-Clang-15 build on ORNL Jenkins CI
```
/var/jenkins/workspace/Kokkos_PR-8476/core/src/Kokkos_View.hpp:485:29: error: missing 'typename' prior to dependent type name 'base_t::reference' [clang-diagnostic-error]
  KOKKOS_FUNCTION constexpr base_t::reference operator()(
                            ^~~~~~~~~~~~~~~~~
                            typename 
```

dalg24 replied:
```suggestion
  KOKKOS_FUNCTION constexpr typename base_t::reference operator()(
```

dalg24 replied:
or better yet
```suggestion
  KOKKOS_FUNCTION constexpr reference_type operator()(
```

dalg24 committed:
Fixup missing typename keyword

Signed-off-by: Damien L-G <dalg24+github@gmail.com>

masterleinad committed:
Avoid macro usage across files

Signed-off-by: Daniel Arndt <arndtd@ornl.gov>

crtrott commented:
Note Stan ran the SYCL workaround version, he is rechecking now

stanmoore1 commented:
I pulled to b6d2339f831ea6f0fd9713cd9ffd6fac5f571b82 and it gives even a little better performance than before.

 | Kokkos Version | VIEW_LEGACY=on | VIEW_LEGACY=off |
 | :------- | :------: | -------: |
 | **4.7.1**    | 38.2   | 34.8   |
 | **PR #8476 b6d2339f831ea6f0fd9713cd9ffd6fac5f571b82**   | 38.3   | 42.3   |

dalg24 commented on core/src/Kokkos_View.hpp:
```diff
@@ -397,6 +402,141 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
+          Kokkos::Impl::IsLayoutRightPadded<
+              typename base_t::layout_type>::value ||
+          std::is_same_v<typename base_t::layout_type, Kokkos::layout_stride>))
+  KOKKOS_FUNCTION constexpr base_t::reference operator()(
```
Uggg missing typename keyword
